Add OpenTracing module

Spikhalskiy opened the pull request:
## What was changed:
This PR adds a new temporal-opentracing module with support for OpenTracing context propagation from the client to workflow and to activity execution.

## Why?
We want to be able to tie all the parts of the workflow execution under the same client parent OpenTracing span for tracing and debugging.

## How it can be used?

See README provided with this module: https://github.com/temporalio/sdk-java/pull/418/files#diff-a87181e0e99e6147518982b133e66674da61be50e9898fe9d1eab9dc91ab2cc6R5

## Closes
#258

Spikhalskiy commented:
There will be a couple more things later today:
1. moving options into optional input parameters
2. failures logging into Span

But this is ready for the main review and feedback.

vitarb commented on temporal-opentracing/src/main/java/io/temporal/opentracing/OpenTracingConfig.java:
```diff
@@ -0,0 +1,125 @@
+package io.temporal.opentracing;
+
+public class OpenTracingConfig {
+  private final String workflowStartOperationNamePrefix;
```
I would put these into the map and make key an enum (e.g. TracingNamePrefix with options like Start, StartWithSignal, etc.). This should make code a bit cleaner both here and on the consumer end.

Spikhalskiy replied:
Good idea, implemented.

vitarb commented on temporal-opentracing/src/main/java/io/temporal/opentracing/internal/OpenTracingContextPropagationUtils.java:
```diff
@@ -0,0 +1,99 @@
+import java.util.Optional;
+
+public class OpenTracingContextPropagationUtils {
+  private static final String tracerHeaderKey = "_tracer-data";
```
Name it `TRACER_HEADER_KEY` since it's a constant.

vitarb commented on temporal-opentracing/src/main/java/io/temporal/opentracing/internal/OpenTracingContextPropagationUtils.java:
```diff
@@ -0,0 +1,99 @@
+import java.util.Map;
+import java.util.Optional;
+
+public class OpenTracingContextPropagationUtils {
```
I would prefer more specific name here, like `OpenTracingContextAccessor` or something along those lines. Util classes tend to become garbage cans very quickly.

Spikhalskiy replied:
The name OpenTracingContextAccessor should be used for a helper class with non-static methods, but I don't mind, renamed.

vitarb commented on temporal-opentracing/src/main/java/io/temporal/opentracing/internal/OpenTracingContextPropagationUtils.java:
```diff
@@ -0,0 +1,99 @@
+public class OpenTracingContextPropagationUtils {
+  private static final String tracerHeaderKey = "_tracer-data";
+
+  public static void writeSpanContextToHeader(
```
Did you consider making these non-static and instantiating this accessor as part of init?

Spikhalskiy replied:
I don't really see too much of a benefit here, this class doesn't really have anything it doesn't take as an input parameter. I can make it a helper if you really prefer, but the method signatures will be totally the same I think.

vitarb replied:
I generally have a preference towards non-static helpers, they are less prone to misuse. But I'm okay leaving it as is.

vitarb commented on temporal-opentracing/src/main/java/io/temporal/opentracing/internal/OpenTracingSpanUtils.java:
```diff
@@ -0,0 +1,150 @@
+    Map<String, String> tags =
+        new HashMap<String, String>() {
+          {
+            put("workflowTag", workflowId);
+            put("parentWorkflowId", parentWorkflowId);
+            put("parentRunId", parentRunId);
```
These tags should probably be a separate enum or a set of constants so it's easy to see what the full set of tags is.

vitarb commented on temporal-opentracing/src/main/java/io/temporal/opentracing/internal/OpenTracingSpanUtils.java:
```diff
@@ -0,0 +1,150 @@
+  public static Tracer.SpanBuilder createWorkflowStartSpan(
+      Tracer tracer, long startTimeMs, String spanOperationName, String workflowId) {
+    Map<String, String> tags =
+        new HashMap<String, String>() {
+          {
+            put("workflowTag", workflowId);
+          }
+        };
```
You can use ImmutableMap.of instead.

Spikhalskiy replied:
Yeah, I thought about it, but I actually don't bring a dependency on guava in this module. I know that `temporal-sdk` has a guava dependency anyway, but I don't see enough reasons to make this module dependant on guava too JUST for this small helper inline map factory method. I will maybe consider if at least something more useful from there will be needed.

vitarb replied:
We have an API dependency on temporal-sdk, which should bring in guava.

Spikhalskiy replied:
That's exactly what I'm saying. You don't suppose to use transitive dependencies in your module, it's actually not a good style. And this is why Gradle has a separation between implementation and API scope. And, by the way, guava is defined as implementation scoped (as it should be). Whenever you use a library directly in your module you should have an own direct dependency defined in the project module. And I don't want to bring the whole guava for a small factory method, it's not a good practice.

mfateev replied:
Agree with @Spikhalskiy on not bringing large direct dependencies without a good reason.

Spikhalskiy replied:
Another suggestion brought guava anyway, so now replaced with ImmutableMap

vitarb commented on temporal-opentracing/src/main/java/io/temporal/opentracing/internal/OpenTracingSpanUtils.java:
```diff
@@ -0,0 +1,150 @@
+      String workflowId,
+      String runId,
+      SpanContext workflowStartSpanContext) {
+    Map<String, String> tags =
+        new HashMap<String, String>() {
+          {
+            put("workflowTag", workflowId);
+            put("runTag", runId);
+          }
+        };
```
Same as above.

vitarb commented on temporal-opentracing/src/main/java/io/temporal/opentracing/internal/OpenTracingSpanUtils.java:
```diff
@@ -0,0 +1,150 @@
+
+  public static Tracer.SpanBuilder createActivityStartSpan(
+      Tracer tracer, long startTimeMs, String spanOperationName, String workflowId, String runId) {
+    Map<String, String> tags =
+        new HashMap<String, String>() {
+          {
+            put("workflowTag", workflowId);
+            put("runTag", runId);
+          }
+        };
```
ditto

vitarb commented on temporal-opentracing/src/main/java/io/temporal/opentracing/internal/OpenTracingSpanUtils.java:
```diff
@@ -0,0 +1,150 @@
+      String runId,
+      SpanContext activityStartSpanContext) {
+    Map<String, String> tags =
+        new HashMap<String, String>() {
+          {
+            put("workflowTag", workflowId);
+            put("runTag", runId);
+          }
+        };
```
ditto

vitarb commented on temporal-opentracing/src/main/java/io/temporal/opentracing/internal/OpenTracingSpanUtils.java:
```diff
@@ -0,0 +1,150 @@
+  public static Tracer.SpanBuilder createWorkflowRunSpan(
+      Tracer tracer,
+      long startTimeMs,
+      String spanOperationName,
```
Since we are already in a context of a span, I would rename this to just `operationName`.

vitarb commented on temporal-opentracing/src/main/java/io/temporal/opentracing/internal/OpenTracingWorkflowInboundCallsInterceptor.java:
```diff
@@ -0,0 +1,68 @@
+        OpenTracingSpanUtils.createWorkflowRunSpan(
+                tracer,
+                Workflow.currentTimeMillis(),
+                openTracingConfig.getWorkflowRunOperationNamePrefix()
+                    + "-"
+                    + Workflow.getInfo().getWorkflowType(),
```
Why do we need to add workflow type at the end? Should we add it into tags instead?

Spikhalskiy replied:
Just WorkflowStart or WorkflowRun as an operation name is very undescriptive and pretty meaningless.

A definition/best practice from [OpenTracing Specification](https://github.com/opentracing/specification/blob/master/specification.md):

> An operation name, a human-readable string which concisely represents the work done by the Span (for example, an RPC method name, a function name, or the name of a subtask or stage within a larger computation). The operation name should be the most general string that identifies a (statistically) interesting class of Span instances. That is, "get_user" is better than "get_user/314159".

Also from [Open Tracing Guide](https://opentracing.io/guides/golang/spans/):

> The reason for choosing general operation names is to allow the tracing systems to do aggregations. For example, Jaeger tracer has an option of emitting metrics for all the traffic going through the application. Having a unique operation name for each span would make the metrics useless. For application or library developers who wish to capture the program arguments in the traces to distinguish them, the recommended solution is to annotate spans with tags or logs. These are discussed in the section below.
> 
> Each operation name should have some semantic meaning to it and should be chosen such that its cardinality is neither too high nor too low. The cardinality of an operation name is defined as follows:
> 
> card(operation X) = total number of spans of operation X
> 
> For example, when using http.url (as operation name) the cardinality would be too high, on the other hand if http.method is chosen as the operation name then the cardinality would be too low.

I think that my current solution is most in line with these guidelines.

vitarb replied:
It actually goes against what you've quoted above, specifically "operation name should be the most general string" as we concatenate specific workflow type to operation.

Spikhalskiy replied:
But you just took the fragment out of the context. The full phrase is "most general string that identifies a (statistically) interesting class of Span instances". And there are more phrases around that explain what does it mean. If you open the first link, after this phrase there is a specific example with good and bad examples:

Operation Name | Guidance
-- | --
get | Too general
get_account/792 | Too specific
get_account | Good, and account_id=792 would make a nice Span tag

You offer to make an operation name EVEN MORE generic thing than this `get` example.
I'm standing on the fact that I spent a good amount of time investigating the different implementation and open tracing guidelines and my current solution is much close to the guidelines.

Spikhalskiy replied:
This is how OpenTelemetry UIs look like:
![c6463a2b-trace-300](https://user-images.githubusercontent.com/532108/113498093-d6597b00-94d7-11eb-8b6f-dd54cfd91e2e.png)
![0_WDqZqICj-V23_1YW](https://user-images.githubusercontent.com/532108/113498095-df4a4c80-94d7-11eb-84c8-bd4355e19d77.png)
![aanuraag_Distributed-Tracing-OpenTelemetry_f1](https://user-images.githubusercontent.com/532108/113498123-1b7dad00-94d8-11eb-959f-dd979c34eb30.jpeg)
To get to tags you need to click on a specific span and open the details. How the operator will work with a sequence of indistinguishable ActivityRun spans?

Also, see OpenTracing implementation for JaxRS2:
https://github.com/opentracing-contrib/java-jaxrs/blob/master/opentracing-jaxrs2/src/main/java/io/opentracing/contrib/jaxrs2/server/OperationNameProvider.java#L46

> Default Microprofile operation name <HTTP method>:<package name>.<Class name>.<method name>
<package name>.<Class name>.<method name> is our ActivityType/WorkflowType

They do provide an alternative strategy
```
  /**
   * Returns HTTP method as operation name
   */
  class HTTPMethodOperationName implements OperationNameProvider {
```
but it's not a default

vitarb commented on temporal-opentracing/src/main/java/io/temporal/opentracing/internal/OpenTracingWorkflowOutboundCallsInterceptor.java:
```diff
@@ -0,0 +1,104 @@
+    return OpenTracingSpanUtils.createChildWorkflowStartSpan(
+        tracer,
+        System.currentTimeMillis(),
+        openTracingConfig.getChildWorkflowStartOperationNamePrefix()
+            + "-"
+            + input.getWorkflowType(),
```
Same as above, should we put workflow type into tags instead of part of the operation name?

vitarb reviewed (changes requested):
Great work Dmitry, I left some comments here, please ping me on slack if you have questions.

Spikhalskiy commented on temporal-opentracing/src/main/java/io/temporal/opentracing/internal/OpenTracingSpanUtils.java:
```diff
@@ -0,0 +1,150 @@
+import javax.annotation.Nullable;
+
+public class OpenTracingSpanUtils {
+  public static Tracer.SpanBuilder createWorkflowStartSpan(
```
Not exactly. Each of these methods, other than just tags, encodes two things - a reference type and if explicit parent span is expected/allowed in this scenario.
I would prefer to leave them the current way and it was my intention to make these methods in such a rigid and specific way, so each scenario is super clean - what are the expected tags, if we expect explicitly passed parent span, and what is the reference type. And we have a single place where we can see the difference between these scenarios.

vitarb replied:
I could see the benefit if these functions were called multiple times, but they are called only once each, which only unnecessarily complicates the code by adding another level of indirection.

Spikhalskiy replied:
I think that this is a good way. It's not a level of indirection, it's a level of incapsulation. Some permutations of parameters are incorrect and these methods provide the required incapsulation to ensure the correctness of permutations of parameter values.

mfateev replied:
I like @Spikhalskiy  approach more. Mostly because all the ugliness of creating tag hashmaps is encapsulated in one place. I think this is about the style which is very subjective. I would leave it as it is just to avoid bikeshedding.

Spikhalskiy replied:
Made it a non-static factory and hid the logic of span operation name creation inside to give an easy entry point for providing extensible strategies in the future if needed.

Spikhalskiy commented:
@vitarb Thanks for your review!

- Addressed PR comments
- Implemented logging of workflow/activity failures
- Reworked and exposed OpenTracingOptions

Ready to be revisited.

orastosk commented on temporal-opentracing/src/main/java/io/temporal/opentracing/internal/OpenTracingContextAccessor.java:
```diff
@@ -0,0 +1,95 @@
+  private static SpanContext deserializeSpanContextFromMap(
+      Map<String, String> serializedSpanContext, Tracer tracer) {
+    HashMapBackedTextMap contextTextMap = new HashMapBackedTextMap(serializedSpanContext);
+    return tracer.extract(Format.Builtin.TEXT_MAP, contextTextMap);
```
you can use io.opentracing.propagation.TextMapAdapter from opentracing-api.jar and remove local HashMapBackedTextMap.
```java
return tracer.extract(Format.Builtin.TEXT_MAP, new TextMapAdapter(serializedSpanContext));
```

Spikhalskiy replied:
That's neat, I replaced this custom `HashMapBackedTextMap` class with standard opentracing classes from `io.opentracing.propagation` package. Thank you for pointing this out!

mfateev commented on temporal-opentracing/src/main/java/io/temporal/opentracing/internal/OpenTracingContextAccessor.java:
```diff
@@ -0,0 +1,70 @@
+  public static void writeSpanContextToHeader(
+      SpanContext spanContext, Header header, Tracer tracer) {
+    Map<String, String> serializedSpanContext = serializeSpanContextToMap(spanContext, tracer);
+    Optional<Payload> payload = DataConverter.getDefaultInstance().toPayload(serializedSpanContext);
```
Shouldn't we use[ tracer.inject ](https://opentracing.io/guides/java/inject-extract/) and `tracer.extract` instead of DataConverter?

Spikhalskiy replied:
Not really. tracer.inject and tracer.extract are working with TEXT_MAPs. So, we convert a span to a TEXT_MAP using tracer.inject inside serializeSpanContextToMap, but after that we need to serialize this map for temporal headers and we use DataConverter for that. tracer.inject is used here, it just happened inside serializeSpanContextToMap.

mfateev replied:
What about using BINARY format to avoid JSON type serialization? You still would need to use DataConverter to create the Payload, but it will use the binary payload as it is.

Spikhalskiy replied:
I don’t mind, I don’t see problems with both approaches. If you prefer Binary - let’s go with Binary. Will delete guava and rework back to Hash Maps then :) Temporal Go SDK uses TEXT_MAP + converter approach though. You sure you want it differently between SDKs?

mfateev replied:
I don't have a strong opinion on this. If you feel that map is more debuggable let's keep it as it is.

Spikhalskiy replied:
Yeah, me too. Let's go with Map. It looks like it's the most used and referenced implementation and it's also in agreement with how things are done in Go SDK.

mfateev commented on temporal-opentracing/src/main/java/io/temporal/opentracing/internal/OpenTracingWorkflowClientCallsInterceptor.java:
```diff
@@ -0,0 +1,74 @@
+  public WorkflowStartOutput start(WorkflowStartInput input) {
+    createAndPassWorkflowStartSpan(
+        input, openTracingOptions.getSpanOperationNamePrefix(SpanOperationType.WORKFLOW_START));
+    return super.start(input);
```
I think we should close this span only after the start gRPC call completes.

Spikhalskiy replied:
It's go-sdk approach though I believe: https://github.com/temporalio/sdk-go/blob/cec615e1aaf798e1b153ba3918b4d3027edde808/internal/internal_workflow_client.go#L199

I'm not against your proposal, we can close in finally. But can you give more motivation or rasoning here? These two solutions lead to the same result for me, maybe I don't see something.

mfateev replied:
My idea was to have a span that includes the start gRPC call latency and possibly retries. I'm not sure if it is useful.

Spikhalskiy replied:
Alright, makes sense, let me do it this way.

mfateev commented on temporal-opentracing/src/main/java/io/temporal/opentracing/internal/OpenTracingWorkflowClientCallsInterceptor.java:
```diff
@@ -0,0 +1,74 @@
+    Tracer.SpanBuilder workflowSpan =
+        createWorkflowStartSpanBuilder(tracer, input, spanOperationNamePrefix);
+    Span span = workflowSpan.start();
+    span.finish();
```
don't close before start call completed

Spikhalskiy replied:
same ^. go-sdk does it the same way too.

mfateev commented on temporal-opentracing/src/main/java/io/temporal/opentracing/internal/OpenTracingWorkflowInboundCallsInterceptor.java:
```diff
@@ -0,0 +1,74 @@
+      OpenTracingSpanUtils.logFail(workflowRunSpan, t);
+      throw t;
+    } finally {
+      workflowRunSpan.finish();
```
This is going to close the workflow span every time a workflow is unloaded from the worker cache. I think it is pretty confusing to see many spans instead of one span for a workflow. I propose the following solution to this:

Any span created inside the workflow should be created inside a `SideEffect` function. It should be serialized using `Tracer.inject` as SideEffect requires its result to be serializable as it is stored as a _marker_ event in the event history. Then it has to be deserialized using `Tracer.extract` before use. 

`workflowRunSpan.finish()` should not be called on exceptions that extend Error, especially `DestroyWorkflowThreadError` which is used to release threads on workflow unloading from memory. The interceptor interface should be extended to give a clear indication if an exception is going to fail workflow or workflow task only.

Spikhalskiy replied:
What do you think about making this a separate change and PR and implement it as an improvement for the current solution?
The current approach is simple, it works and it could be good enough for the majority of cases.

Also, this idea looks very brittle for me and I think it will not work. The first problem that I see is actually in the root of the proposal:
> This is going to close the workflow span every time a workflow is unloaded from the worker cache.
> workflowRunSpan.finish() should not be called on exceptions that extend Error, especially DestroyWorkflowThreadError which is used to release threads on workflow unloading from memory.

We should call finish in all these cases. Otherwise, imagine a workflow started today and finishing a month from now. A workflow run span of this workflow will not be finished till a month from now? It will not be reported to the server in that case and we will not have any trace.

We should actually wrap every "awakening" as a separate span to keep spans published without huge delays.

mfateev replied:
OK. Let's keep your approach. I didn't realize that children of non-closed span are not reported. So it is going to be useless for long running workflows.

mfateev commented on temporal-opentracing/src/main/java/io/temporal/opentracing/internal/OpenTracingWorkflowOutboundCallsInterceptor.java:
```diff
@@ -0,0 +1,107 @@
+
+  @Override
+  public <R> ActivityOutput<R> executeActivity(ActivityInput<R> input) {
+    createAndPassActivityStartSpan(input.getActivityName(), input.getHeader());
```
I would change it to a span that wraps activity execution, not only its invocation. This comment applies to all other wrapped APIs.

Spikhalskiy replied:
For wrapping the activity execution with have a RunActivity span. It wraps the execution. This is a StartActivity span. It just basically shows that the activity was triggered/scheduled. I think we should close the span either immediately or after invocation to show in the trace that it happened. It can be especially important for WorkflowStart span of long-living activity. You don't want to wait till the end of a workflow to publish a root span.

Spikhalskiy commented on temporal-opentracing/src/main/java/io/temporal/opentracing/SpanOperationType.java:
```diff
@@ -0,0 +1,39 @@
+
+public enum SpanOperationType {
+  WORKFLOW_START("StartWorkflow"),
+  WORKFLOW_START_WITH_SIGNAL("SignalWithStartWorkflow"),
```
Alright. I thoughtt that it should be ok to keep SpanTypes and their prefix a little bit different and kind of like the current version, but if you think it's better to make enum names match prefixes - will do it.

Spikhalskiy commented on temporal-opentracing/src/main/java/io/temporal/opentracing/SpanOperationType.java:
```diff
@@ -0,0 +1,39 @@
+
+  private String defaultPrefix;
+
+  SpanOperationType(String defaultPrefix) {
```
Any Enum constructor in java is private, it can't be anything else.

Spikhalskiy commented on temporal-opentracing/src/main/java/io/temporal/opentracing/internal/OpenTracingActivityInboundCallsInterceptor.java:
```diff
@@ -0,0 +1,81 @@
+
+  @Override
+  public ActivityOutput execute(ActivityInput input) {
+    Tracer tracer = GlobalTracer.get();
```
From what I know - GlobalTracer is actually the conventional way to tell all components which tracer to use. But if we really want such flexibility to be able to provide something else to temporal - we can expose AtomicReference<Tracer> in OpenTracingOptions. It's not pretty, but we should pass a container with a tracer, not the tracer itself to keep a way to replace the Tracer later (like we can do with replacing the tracer in GlobalTracer).

Spikhalskiy commented on temporal-opentracing/src/main/java/io/temporal/opentracing/internal/OpenTracingContextAccessor.java:
```diff
@@ -0,0 +1,70 @@
+
+package io.temporal.opentracing.internal;
+
+import com.google.gson.reflect.TypeToken;
```
Replaced with TypeToken from guava.
